### PR TITLE
Add transparency and layering for overlays

### PR DESCRIPTION
**Pull request by zzmp**
Commits d14ce8a..a790b09 update the geometry cache to:
- Allow passing of an explicit ShapePipeline into instanced draw calls, in order to add buffers.
- Supply a normal buffer to draw calls, and fix those for instanced shapes.

Commits 127f969..60f3ee5 implement various fixes necessary for overlays to work.

Commits 399fe95..53bd595 update the overlay and emissive shaders and pipelines (through their plumbers) and update all appropriate overlays to render through them.

The effect of all of this is to allow overlays to be layered (if `drawInFront` is set) and transparent (if `alpha != 0.1`).

### CODE REVIEW

**Comment by zzmp** [comment]
This is the script I used to test this - it may be helpful. To use it, go to `0,0,0`, step back a meter or two, and call `test(TYPE)`, where `TYPE` is the overlay you'd like to test: text3d, image3d, web3d, circle3d, line3d, rectangle3d, cube, or sphere. After viewing a test, call `teardown()` to delete the overlays before beginning a new test.
```
// Tests are defined as
// arch: the archetypal overlay object
// prop: a single extra boolean property of the overlay to test
// func: a function to run after generating the definition from the arch, but before creating the overlay
//       it is passed in the generated definition
// e.g., image3d will generate 8 tests, all of which will have url set, 4 of which will be emissive: true
var tests = {
  text3d: { arch: { text: 'Hello, world!' } },
  image3d: {
    prop: 'emissive',
    arch: { url: 'https://lh4.ggpht.com/wKrDLLmmxjfRG2-E-k5L5BUuHWpCOe4lWRF7oVs1Gzdn5e5yvr8fj-ORTlBF43U47yI=w300' },
  },
  // grid must be tested separately
  circle3d: { arch: { solid: true } },
  line3d: {
    prop: 'dashed',
    func: function(obj) {
      obj.start = obj.position;
      delete obj.position;
      obj.end = { x: obj.start.x + 1, y: obj.start.y, z: obj.start.z };
    },
    arch: { color: { red: 255, blue: 0, green: 0 } },
  },
  rectangle3d: { arch: { solid: true, dimensions: { width: 1, height: 1 } } },
  web3d: {
    arch: { url: 'https://lh4.ggpht.com/wKrDLLmmxjfRG2-E-k5L5BUuHWpCOe4lWRF7oVs1Gzdn5e5yvr8fj-ORTlBF43U47yI=w300' },
  },
  cube: {
    prop: 'solid',
    arch: { color: { red: 255, blue: 0, green: 0 } }
  },
  sphere: {
    prop: 'solid',
    arch: { color: { red: 255, blue: 0, green: 0 } },
  },
};

var overlays = [];

function position(i) {
  return { x: i%4, y: Math.floor(i/4), z: 0 };
}

function teardown() {
  overlays.forEach(Overlays.deleteOverlay);
  overlays = [];
}

function test(type) {
  var prop = tests[type].prop;
  var func = tests[type].func;
  var arch = tests[type].arch;
  var max = prop ? 8 : 4;

  for (var i = 0; i < max; i++) {
    var obj = { position: position(i) };
    if (arch) for (var p in arch) obj[p] = arch[p];

    // drawInFront
    obj.drawInFront = ( i < max/2 );

    // alpha
    obj.alpha = ( i % 2 == 0 ) ? 1.0 : 0.7;

    // extras
    if (prop) obj[prop] = (i % 4 < 2);

    if (func) func(obj);

    overlays.push(Overlays.addOverlay(type, obj));
  }
}
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79d00cc7808e6dce99a0602a3d4e15e949245d14 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by zzmp** [comment]
I believe an alpha is set by default on overlays - do we want to keep this behavior?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 53bd595b776f9903f9fc47d2de2a101bc6826b1a available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cb06d4a9e451ed62671a62f13d648fc59478179f available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by zzmp** [comment]
If an instanced overlay is rendered in the DrawOpaqueDeferred pass, as well as a RenderableEntity (i.e. spheres or cubes), whichever is rendered "first" will set the pipeline for all named calls of that type for that batch. This means that it will sometimes use the geometry cache's simple program (simple.slv->simple_textured.slf), and sometimes use the model's program (model.slv->model.slf). This may hold true for transparents as well - I'm not sure if RenderableEntities can be made transparent.

This should be fixed before this is merged. My best idea to fix it right now is to have named calls be further sorted by their passed in pipelines.

**Comment by zzmp** [comment]
@cozza13 I want to put one more commit on here prior to QA testing to deal with the above comment.

**Comment by zzmp** [comment]
I thought 44d297e would fix it, but there is some disparity between the instanced shapes being rendered by the opaque model pipeline and the opaque overlay pipeline. The overlay pipeline has specular highlights, while the model pipeline looks very matte.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 44d297e7e24674ce5034b6ccb993aed37df4a906 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by zzmp** [comment]
This is still not working because the lightBuffer._direction.w in the overlay shaders is not being properly set. I explicitly set the ambient intensity to 1.0 prior to a render call, and it still showed as less than that in the gpu. I am not sure if it is being overridden somewhere or not. It is the last uniform set aside from named call plumbing.

I'll continue to look into it tomorrow.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 743c4b9b4ba866f217c3db17a600b4b99392ade4 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0a9c299edb52c8d4e3eae4509a8883fe39fb02e available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9132f397316c07c7148ec845fda870e6fcfb950 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by Triplelexx** [comment]
I tried to test this at 0,0,0 but could not seem to get anything to appear.
Maybe I misunderstood something. Copied test script with additions, such as.
`Controller.mousePressEvent.connect(test(line3d));
Script.scriptEnding.connect(teardown);`

**Comment by zzmp** [comment]
@Triplelexx you'll need to pass the type as a string: `test('line3d')`. I updated the instructions above to reflect that, sorry it wasn't included before.

Also, your connect handler will fire when you define it. If you use it, you'll want something like `Controller.mousePressEvent.connect(function() { test('line3d'); });` to avoid invoking `test` right away.

It might be easier to run this on-the-fly instead of loading it as a script, so that you can iterate through all of the test cases at your own pace.

**Comment by Triplelexx** [comment]
I'm not sure I understand what you mean by run on-the-fly, my scripting experience is still rather limited. I'll see if I can figure it out though.
Testing again.

File: libraries/render-utils/src/GeometryCache.h [hunk 1]
```diff
@@ -150,45 +150,55 @@ class GeometryCache : public Dependency {
...
     }
     
-    void renderWireSphereInstance(gpu::Batch& batch, const glm::vec4& color);
-    void renderWireSphereInstance(gpu::Batch& batch, const glm::vec3& color) {
-        renderWireSphereInstance(batch, glm::vec4(color, 1.0));
+    void renderWireSphereInstance(gpu::Batch& batch, const glm::vec4& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline);
+    void renderWireSphereInstance(gpu::Batch& batch, const glm::vec3& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline) {
+        renderWireSphereInstance(batch, glm::vec4(color, 1.0), pipeline);
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.h]: need the f for the 1.0

File: libraries/render-utils/src/GeometryCache.h [hunk 2]
```diff
@@ -150,45 +150,55 @@ class GeometryCache : public Dependency {
...
-        renderSolidSphereInstance(batch, glm::vec4(color, 1.0));
+    // Static (instanced) geometry
+    void renderShapeInstances(gpu::Batch& batch, Shape shape, size_t count, gpu::BufferPointer& colorBuffer);
+    void renderWireShapeInstances(gpu::Batch& batch, Shape shape, size_t count, gpu::BufferPointer& colorBuffer);
+
+    void renderSolidSphereInstance(gpu::Batch& batch, const glm::vec4& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline);
+    void renderSolidSphereInstance(gpu::Batch& batch, const glm::vec3& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline) {
+        renderSolidSphereInstance(batch, glm::vec4(color, 1.0), pipeline);
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.h]: need the f for the 1.0

File: libraries/render-utils/src/GeometryCache.h [hunk 3]
```diff
@@ -150,45 +150,55 @@ class GeometryCache : public Dependency {
...
     }
     
-    void renderSolidCubeInstance(gpu::Batch& batch, const glm::vec4& color);
-    void renderSolidCubeInstance(gpu::Batch& batch, const glm::vec3& color) {
-        renderSolidCubeInstance(batch, glm::vec4(color, 1.0));
+    void renderSolidCubeInstance(gpu::Batch& batch, const glm::vec4& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline);
+    void renderSolidCubeInstance(gpu::Batch& batch, const glm::vec3& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline) {
+        renderSolidCubeInstance(batch, glm::vec4(color, 1.0), pipeline);
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.h]: need the f for the 1.0

File: libraries/render-utils/src/GeometryCache.h [hunk 4]
```diff
@@ -150,45 +150,55 @@ class GeometryCache : public Dependency {
...
     }
     
-    void renderWireCubeInstance(gpu::Batch& batch, const glm::vec4& color);
-    void renderWireCubeInstance(gpu::Batch& batch, const glm::vec3& color) {
-        renderWireCubeInstance(batch, glm::vec4(color, 1.0));
+    void renderWireCubeInstance(gpu::Batch& batch, const glm::vec4& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline);
+    void renderWireCubeInstance(gpu::Batch& batch, const glm::vec3& color,
+                                    const render::ShapePipelinePointer& pipeline = _simplePipeline) {
+        renderWireCubeInstance(batch, glm::vec4(color, 1.0), pipeline);
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.h]: need the f for the 1.0

File: libraries/render-utils/src/overlay3D_emissive.slf [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+out vec4 _fragColor;
+
+void main(void) {
+    vec4 diffuse = texture(originalTexture, _texCoord0);
+
+    if (diffuse.a <= 0.1) {
+        discard;
+    }
+    _fragColor = vec4(diffuse.rgb * _color, diffuse.a);
```
Comment (samcake) [inline libraries/render-utils/src/overlay3D_emissive.slf]: Probably need gamma correction here too ?

File: libraries/render/src/render/ShapePipeline.h [hunk 6]
```diff
@@ -28,7 +28,9 @@ class ShapeKey {
         SKINNED,
         STEREO,
         DEPTH_ONLY,
+        DEPTH_BIAS,
         WIREFRAME,
+        CULL,
```
Comment (samcake) [inline libraries/render/src/render/ShapePipeline.h]: As we just discussed, the name would be NO_CULL_FACE

File: libraries/render/src/render/ShapePipeline.h [hunk 7]
```diff
@@ -39,12 +41,12 @@ class ShapeKey {
 
     Flags _flags;
 
-    ShapeKey() : _flags{0} {}
+    ShapeKey() : _flags{0} { _flags.set(CULL); }
```
Comment (samcake) [inline libraries/render/src/render/ShapePipeline.h]: And so NO_CULL_FACE is false by default so no need for this one

File: libraries/render/src/render/ShapePipeline.h [hunk 8]
```diff
@@ -39,12 +41,12 @@ class ShapeKey {
...
     ShapeKey(const Flags& flags) : _flags{flags} {}
 
     class Builder {
     public:
-        Builder() {}
+        Builder() { _flags.set(CULL); }
```
Comment (samcake) [inline libraries/render/src/render/ShapePipeline.h]: Same as above

**Comment by samcake** [comment]
Just reviewed the code and pretty happy about it.
I will tag as Developer approved once the comments are fixed

**Comment by Triplelexx** [comment]
seems to be working so far for me as well, just getting some unexpected differences in color due to lighting.

**Comment by ctrlaltdavid** [comment]
Successfully tested with edit.js: you can now see all the sizing handles through both cube/sphere entities and model entities whereas you couldn't before.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a7bafed61b968a6ad9f9c83b9604e68ba69295b9 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by cozza13** [comment]
Fixes for https://github.com/highfidelity/hifi/pull/7028
https://app.asana.com/0/26225263936266/29389937706532
https://app.asana.com/0/26225263936266/76133442688381

**Comment by samcake** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: a7bafed61b968a6ad9f9c83b9604e68ba69295b9 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 281ba3c3eac89f35e7e9f9b5be842023399e11a3 available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by zzmp** [comment]
@samcake found it to be crashing on shadows:
1. Enable shadows
2. Run edit.js
3. Click on something

Seems to be overlay3d stuff crashing it. Crashes in RenderShadowTask::run.

**Comment by zzmp** [comment]
#### Intents
- Add layering for overlays (i.e. they should render *in front* of other things if the `drawInFront` property is set).
- Add transparency for overlays.

#### Testing
- Test the different types of overlays through scripting, using the script and instructions in the first comment
- Test edit.js, and make sure overlays appear in a sensical way
- Make sure to test both with and without shadows (`Render.RenderShadowTask.enabled = true` or `false`)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 63d87550b1a7f003822345834632c5486a3613bc available at: https://gist.github.com/7cf8f7d2d449b2f3bfb2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7028.dmg

**Comment by samcake** [comment]
ok tested successfully

**Comment by nbq** [comment]
Testing now

**Comment by nbq** [comment]
Using "Render.RenderShadowTask.enabled = true" I have tested:
text3d - worked
image3d - worked 
web3d - worked (lags a little after stopping script)
circle3d - worked 
line3d - worked
cube - worked
sphere - worked

Using "Render.RenderShadowTask.enabled = false" I have tested:
text3d - worked fine
image3d - worked
web3d - works (still lags a little)
circle3d - works 
line3d - works
cube - works
sphere - works

After testing all of these I got the "Due to the complexity of the content, the lod has been adjusted" message, which I never get as I have a GTX 970, 16gig ram etc. I it tends to get worse after the web3d tests.

**Comment by ZappoMan** [comment]
@nbq - this looks like it should be "QA Approved"? Is there a reason you didn't mark it as such?

**Comment by cozza13** [comment]
@ZappoMan , at the moment I mark them QA approved after reviewing any comments

**Pull request merged**